compact.go

```diff
@@ -133,37 +145,89 @@ func (c *compactor) Plan() ([][]string, error) {
 		return dms[i].meta.MinTime < dms[j].meta.MinTime
 	})
 
-	if len(dms) == 0 {
+	if len(dms) <= 1 {
 		return nil, nil
 	}
 
-	sliceDirs := func(i, j int) [][]string {
+	sliceDirs := func(dms []dirMeta) [][]string {
+		if len(dms) == 0 {
+			return nil
+		}
 		var res []string
-		for k := i; k < j; k++ {
-			res = append(res, dms[k].dir)
+		for _, dm := range dms {
+			res = append(res, dm.dir)
 		}
 		return [][]string{res}
 	}
 
-	// Then we care about compacting multiple blocks, starting with the oldest.
-	for i := 0; i < len(dms)-compactionBlocksLen+1; i++ {
-		if c.match(dms[i : i+3]) {
-			return sliceDirs(i, i+compactionBlocksLen), nil
+	return sliceDirs(c.selectDirs(dms)), nil
+}
+
+func (c *compactor) selectDirs(ds []dirMeta) []dirMeta {
+	// The way tp skip compaction is to not have blockRanges.
+	if len(c.opts.blockRanges) == 1 {
+		return nil
+	}
+
+	return selectRecurse(ds, c.opts.blockRanges)
+}
+
+func selectRecurse(dms []dirMeta, intervals []int64) []dirMeta {
+	if len(intervals) == 0 {
+		return dms
+	}
+
+	// Get the blocks by the max interval
+	blocks := splitByRange(dms, intervals[len(intervals)-1])
+	dirs := []dirMeta{}
+	for i := len(blocks) - 1; i >= 0; i-- {
+		// We need to choose the oldest blocks to compact. If there are a couple of blocks in
+		// the largest interval, we should compact those first.
+		if len(blocks[i]) > 1 {
+			dirs = blocks[i]
+			break
 		}
 	}
 
-	return nil, nil
+	// If there are too many blocks, see if a smaller interval will catch them.
+	// i.e, if we have 0-20, 60-80, 80-100; all fall under 0-240, but we'd rather compact 60-100
+	// than all at once.
+	if len(dirs) > 2 {
+		smallerDirs := selectRecurse(dirs, intervals[:len(intervals)-1])
+		if len(smallerDirs) > 1 {
+			return smallerDirs
+		}
+	}
+
+	return dirs
 }
 
-func (c *compactor) match(dirs []dirMeta) bool {
-	g := dirs[0].meta.Compaction.Generation
+func splitByRange(ds []dirMeta, tr int64) [][]dirMeta {
+	splitDirs := [][]dirMeta{}
+	t0 := ds[0].meta.MinTime - ds[0].meta.MinTime%tr
+	dirs := []dirMeta{}
 
-	for _, d := range dirs {
-		if d.meta.Compaction.Generation != g {
-			return false
+	for _, dir := range ds {
+		if intervalContains(t0, t0+tr, dir.meta.MinTime) && intervalContains(t0, t0+tr, dir.meta.MaxTime) {
+			dirs = append(dirs, dir)
+			continue
+		}
+
+		if dir.meta.MinTime >= t0+tr {
+			splitDirs = append(splitDirs, dirs)
+			dirs = []dirMeta{}
+			t0 = dir.meta.MinTime - dir.meta.MinTime%tr
+			if intervalContains(t0, t0+tr, dir.meta.MinTime) && intervalContains(t0, t0+tr, dir.meta.MaxTime) {
+				dirs = append(dirs, dir)
+			}
 		}
```

Comment:
Repeating initialization and the  condition at different indentation levels makes is hard to grasp.

```go
var splitDirs [][]dirMeta

for i := 0; i < len(ds); {
    var group []dirMeta
    // Compute start of aligned time range of size tr closest to the current block's start.
    t0 := ds[i].meta.MinTime - (ds[i].meta.MinTime % tr)

    // Add all dirs to the current group that are within [t0, t0+tr].
    for ;i < len(ds); i++ {
        if ds[i].meta.MinTime < t0 || ds[i].meta.MaxTime > t0+tr {
            break
        }
        group = append(group, ds[i])
    }

    if len(group) > 0 {
        splitDirs = append(splitDirs, group)
    }
}

return splitDirs
```

Haven't verified this one, but should roughly do the same.